### ISSUE TITLE
BayesNet forward_backward returns incorrect results

### ISSUE THREAD

**Issue by jmschrei** [jmschrei/pomegranate#5]
The backward step in the sum-product algorithm is producing incorrect results. I'm on the case.

**Comment by jmschrei** [comment]
This seems to be because I've implemented a mutant algorithm between the junction-tree and the flooding schedule, which will give correct results given certain graph structures, but incorrect results given other structures. I will be rewriting this code shortly.

**Comment by jmschrei** [comment]
This has been solved by my latest commit.

To summarize:
I had implemented an incorrect version of the belief propagation algorithm, which corresponded to an incorrect parsing of the inference tree in factor graph belief propagation. Some values would be calculated correctly, but this was largely dependent on graph structure.

In addition to being incorrect, this code was also hacky.

I have since implemented factor graphs explicitly (such that users can make their own factor graphs if they want!), which act as a back end for Bayes nets and Markov nets when I add them. Belief propagation is now done by converting the Bayesian network into a factor graph automatically and using a flooding message passing schedule in loopy belief propagation on the factor graph directly. 

Loopy belief propagation is known to converge to approximate but good results, and sometimes to perfect results, dependent on the graph structure. I am converging to the same results as other proven-correct models which use loopy belief propagation, but I am not converging to the exact marginals you would get if you ran variable elimination for each node. This is to be expected.

I have significantly altered the code base around Bayesian networks in order to force people to realize that the code they were using before was returning incorrect results.

Apologies for the inconvenience.
